flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/HistogramBasedWatermarkEmitter.java

```diff
@@ -0,0 +1,263 @@
+
+	private static final long serialVersionUID = 1L;
+
+	private boolean testing;
+
+	/**
+	 * A map holding the histogram. Currently we assume
+	 * buckets of 2^10 milliseconds, which is roughly a second.
+	 */
+	private final Map<Integer, Integer> histBuckets = new TreeMap<>();
+
+	/**
+	 * The allowed lateness, i.e. the amount of time the emitted watermark
+	 * will lag behind the maximum timestamp seen so far.
+	 */
+	private long allowedLateness = 0;
+
+	/** The maximum timestamp seen so far (in event-time). */
+	private long currentMaxSeenTimestamp = Long.MIN_VALUE;
+
+	/** A flag indicating if we are in a sampling period or not. */
+	private boolean inSampling = false;
+
+	/** The total number of events seen in the present sampling period. */
+	private int noOfEvents = 0;
+
+	// User-defined parameters.
+
+	private final double percentile;
+	private final long nonSamplingInterval;
+	private final long samplingPeriod;
+
+	/**
+	 * A best-effort periodic watermark emitter. For more details see {@link HistogramBasedWatermarkEmitter}.
+	 * @param samplingPeriod
+	 *             the duration of the sampling.
+	 * @param nonSamplingPeriod
+	 * 				the interval between the end of a sampling and the
+	 * 				beginning of the next one.
+	 * @param percent
+	 * 				the percentage of late and non-late elements in the stream
+	 * 				to cover as non-late.
+	 * */
+	public HistogramBasedWatermarkEmitter(Time samplingPeriod,
+										Time nonSamplingPeriod,
+										double percent) {
+
+		if (samplingPeriod == null || nonSamplingPeriod == null) {
+			throw new RuntimeException("Tried to set: " +
+				"samplingPeriod=" + samplingPeriod + ", " +
+				"interSamplingInterval=" + nonSamplingPeriod + ".\n" +
+				"These parameters cannot be null.");
+		}
+
+		if (percent < 0) {
+			throw new RuntimeException("The percentage of data covered cannot be negative.");
+		}
+
+		this.samplingPeriod = samplingPeriod.toMilliseconds();
+		this.nonSamplingInterval = nonSamplingPeriod.toMilliseconds();
+		this.percentile = percent;
+		restoreToInit();
+	}
+
+	public long getSamplingPeriodDurationInMillis() {
+		return samplingPeriod;
+	}
+
+	public long getNonSamplingIntervalDurationInMillis() {
+		return nonSamplingInterval;
+	}
+
+	public double getCoveragePercentile() {
+		return this.percentile;
+	}
+
+	/**
+	 * Extracts the timestamp from a record in the stream.
+	 * @param element The element from which to extract the timestamp.
+	 */
+	public abstract long extractTimestamp(T element);
+
+	@Override
+	public void open(Configuration parameters) throws Exception {
+		super.open(parameters);
+		inSampling = true;
+		setNextSamplingTimer(true);
+	}
+
+	@Override
+	public Watermark getCurrentWatermark() {
+
+		// we do not risk overflow because we always subtract from the currentMaxSeenTimestamp
+		// we do not risk underflow because the allowed lateness will always be smaller than the
+		// currentMaxSeenTimestamp and bigger than Long.MIN_VALUE
```

Comment:
I don't understand how the condition bigger than `Long.MIN_VALUE` guards against overflows? What if you calculate `1 - (Long.MIN_VALUE - 1)`, then you would have an overflow.